imagebuildah/stage_executor.go

```diff
@@ -1175,6 +1199,13 @@ func (s *StageExecutor) Execute(ctx context.Context, base string) (imgID string,
 		// image, run the step so that we can check if the result
 		// matches a cache.
 		if cacheID == "" {
+			// Note: In theory we should never ever hit this branch now
+			// since theoretically only `COPY` and `ADD` should  fall in
+			// this branch but we already process them with execption cases
+			// above to save cycles. But still leave this branch here so if
+			// if we missed anything then we still end up giving our best
+			// effort to find a cache hit.
+
 			// Process the instruction directly.
 			if err = ib.Run(step, s, noRunsRemaining); err != nil {
```

Comment:
If we have a cache miss for an ADD or COPY instruction, are we copying the content again?

Reply:
@nalind Yes if miss `cache` hit for ADD or COPY content gets copied again.I can add a test to verify this in CI as well.

Reply:
I am even ok to drop this branch entirely as discussed here: https://github.com/containers/buildah/pull/4128#discussion_r926699466

Reply:
Copying the content twice can be noticeably time-consuming, so it's something we should try to avoid.

Reply:
Actually this might a better condition for this branch `if cacheID == "" && step.Command != "copy" && step.Command != "add" {
`

Reply:
(Personal style, this is up to Buildah’s maintainers, not me:)

I’d find a copy&pasted `step.Command != "..."` hard to maintain. I would very much worry that it is going to get out of sync.

Using a named variable like `canMatchCacheOnlyAfterRun` ensures exact consistency between the code paths, and allows using IDE’s functions to find where it is set and used.

(Also, long comments are _great_, but they still can get out of sync. So, specifically named variables that _precisely_ describe the decision they make, and contain “obviously correct” logic that doesn’t even need a comment, are preferable, if they happen to be possible. But that’s _totally_ a personal style thing.)

Reply:
@mtrmac SGTM. One the core logic in PR gets approved I'll push again with this suggestion.

Reply:
resolved.